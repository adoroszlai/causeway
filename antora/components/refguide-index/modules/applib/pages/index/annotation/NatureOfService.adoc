= NatureOfService _(enum)_
:Notice: Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file to you under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at. http://www.apache.org/licenses/LICENSE-2.0 . Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR  CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

The different sorts of domain services recognized by Causeway, as specified in _DomainService#nature()_

== API

[source,java]
.NatureOfService.java
----
enum NatureOfService {
  WEB_UI     // <.>
  WEB_API     // <.>
  BOTH     // <.>
  boolean isBoth()     // <.>
  boolean isWebUi()     // <.>
  boolean isWebApi()     // <.>
}
----

<.> xref:#WEB_UI[WEB_UI]
+
--
The service's actions appear only in the menu bar human-usable Web UIs (such as Wicket viewer). They do _not_ appear in any REST or GraphQL APIs.
--
<.> xref:#WEB_API[WEB_API]
+
--
The service's actions should only be visible in the Web API exposed by the Restful Objects viewer and the GraphQL viewer. They do _not_ appear in any human-usable Web UIs (such as Wicket viewer)
--
<.> xref:#BOTH[BOTH]
+
--
The service's actions appear in the menu bar of Web UIs (such as Wicket viewer), and also appear in the Web APIs (Restful Objects viewer and GraphQL viewer).
--
<.> xref:#isBoth_[isBoth()]
+
--
Whether a service contributes its actions to both human-usable Web UIs and the Web APIs.
--
<.> xref:#isWebUi_[isWebUi()]
+
--
Whether a service contributes its actions exclusively to human-usable Web UIs.
--
<.> xref:#isWebApi_[isWebApi()]
+
--
Whether a service contributes its actions to Web APIs (REST and GraphQL)
--

== Members

[#WEB_UI]
=== WEB_UI

The service's actions appear only in the menu bar human-usable Web UIs (such as Wicket viewer). They do _not_ appear in any REST or GraphQL APIs.

[#WEB_API]
=== WEB_API

The service's actions should only be visible in the Web API exposed by the Restful Objects viewer and the GraphQL viewer. They do _not_ appear in any human-usable Web UIs (such as Wicket viewer)

[#BOTH]
=== BOTH

The service's actions appear in the menu bar of Web UIs (such as Wicket viewer), and also appear in the Web APIs (Restful Objects viewer and GraphQL viewer).

[#isBoth_]
=== isBoth()

Whether a service contributes its actions to both human-usable Web UIs and the Web APIs.

[#isWebUi_]
=== isWebUi()

Whether a service contributes its actions exclusively to human-usable Web UIs.

[#isWebApi_]
=== isWebApi()

Whether a service contributes its actions to Web APIs (REST and GraphQL)
