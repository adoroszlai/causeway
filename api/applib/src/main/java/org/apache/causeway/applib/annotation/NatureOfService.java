/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.applib.annotation;

/**
 * The different sorts of domain services recognized by Causeway, as specified in {@link DomainService#nature()}
 *
 * @since 1.x {@index}
 */
public enum NatureOfService {
    /**
     * The service's actions appear only in the menu bar human-usable Web UIs (such as Wicket viewer).
     * They do <i>not</i> appear in any REST or GraphQL APIs.
     */
    WEB_UI,

    /**
     * The service's actions should only be visible in the Web API exposed by the Restful Objects viewer and the
     * GraphQL viewer.  They do <i>not</i> appear in any human-usable Web UIs (such as Wicket viewer)
     */
    WEB_API,

    /**
     * The service's actions appear in the menu bar of Web UIs (such as Wicket viewer), and also appear in the
     * Web APIs (Restful Objects viewer and GraphQL viewer).
     *
     * @apiNote
     * Contributing actions to the 'viewer' implies, that these must also be exposed to the REST API,
     * simply because alternative viewers might be solely based on the provided REST end-points.
     */
    BOTH,
    ;

    // -- BASIC PREDICATES

    /**
     * Whether a service contributes its actions to both human-usable Web UIs and the Web APIs.
     *
     * @see NatureOfService#BOTH
     */
    public boolean isBoth() {
        return this == BOTH;
    }

    /**
     * Whether a service contributes its actions exclusively to human-usable Web UIs.
     * @see NatureOfService#WEB_UI
     */
    public boolean isWebUi() {
        return this == WEB_UI || this == BOTH;
    }

    /**
     * Whether a service contributes its actions to Web APIs (REST and GraphQL)
     * @see NatureOfService#WEB_API
     */
    public boolean isWebApi() {
        return this == WEB_API || this == BOTH;
    }


}
