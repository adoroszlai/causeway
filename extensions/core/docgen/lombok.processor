The presence of this file activates a profile (defined in bom/pom.xml) that configures maven-compiler-plugin to run the lombok annotation processor.

This is a workaround; the compiler ought to pick up all available processors on the classpath anyway; unknown why it does not, but is probably a badly formed pom.xml somewhere causing dependencies to be ignored.

The one place where this file is _not_ required is with the JDO modules that declare @PersistenceCapable entities.
Such modules need both lombok AND the DN annotation processor (datanucleus-jdo-query) that generates the Qxxx query classes.
For some reason, with these JDO modules, the default behaviour (picking up all available processors) does seem to work.

