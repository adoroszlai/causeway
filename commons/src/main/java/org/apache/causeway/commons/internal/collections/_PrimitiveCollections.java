/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.commons.internal.collections;

import java.util.Iterator;
import java.util.OptionalInt;
import java.util.PrimitiveIterator;
import java.util.stream.IntStream;

import lombok.experimental.UtilityClass;

/**
 * <h1>- internal use only -</h1>
 * Primitive int list implementation.
 * <p>
 * <b>WARNING</b>: Do <b>NOT</b> use any of the classes provided by this package! <br/>
 * These may be changed or removed without notice!
 *
 * @since 2.0
 */
@UtilityClass
public class _PrimitiveCollections {

    /**
     * Primitive int list implementation. Can also operate as a set,
     * if {@link IntList#addUnique(int)} is used over {@link IntList#add(int)}.
     * @implNote not thread-safe
     */
    public static class IntList implements Iterable<Integer> {

        private static final int DEFAULT_INITIAL_CAPACITY = 8;

        private int[] buf;
        private int size = 0;

        public int size() {
            return size;
        }

        public boolean isEmpty() {
            return size==0;
        }
        public boolean isNotEmpty() {
            return size!=0;
        }

        public IntList addUnique(final int v) {
            if(!contains(v)) return add(v);
            return this;
        }

        public IntList add(final int v) {
            if(buf==null) {
                this.buf = new int[DEFAULT_INITIAL_CAPACITY];
            } else if(size==buf.length) {
                var old = buf;
                this.buf = new int[buf.length * 2];
                System.arraycopy(old, 0, buf, 0, size);
            }
            buf[size++] = v;
            return this;
        }

        public int get(final int index) {
            if(index<0
                    || index>=size) {
                throw new IndexOutOfBoundsException(index);
            }
            return buf[index];
        }

        public OptionalInt indexOf(final int v) {
            if(isEmpty()) return OptionalInt.empty();
            for (int i = 0; i < size; i++) {
                if(buf[i]==v) return OptionalInt.of(i);
            }
            return OptionalInt.empty();
        }

        public boolean contains(final int v) {
            if(isEmpty()) return false;
            for (int i = 0; i < size; i++) {
                if(buf[i]==v) return true;
            }
            return false;
        }

        public IntStream stream() {
            return IntStream.of(toArray());
        }

        /**
         * @return a new array containing all the int(s) of this list
         */
        public int[] toArray() {
            var result = new int[size];
            if(!isEmpty()) {
                System.arraycopy(buf, 0, result, 0, size);
            }
            return result;
        }

        @Override
        public Iterator<Integer> iterator() {
            var defensiveCopy = toArray();
            return new PrimitiveIterator.OfInt() {
                int index = 0;
                @Override public boolean hasNext() {
                    return index<defensiveCopy.length;
                }
                @Override public int nextInt() {
                    return defensiveCopy[index++];
                }
            };
        }

    }

}
