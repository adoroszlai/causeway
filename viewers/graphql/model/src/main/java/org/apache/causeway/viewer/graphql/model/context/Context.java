/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.viewer.graphql.model.context;

import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import graphql.schema.GraphQLCodeRegistry;

import org.apache.causeway.viewer.graphql.model.domain.common.query.GqlvDomainObject;
import org.apache.causeway.viewer.graphql.model.domain.common.query.GqlvDomainService;

import org.springframework.stereotype.Component;

import org.apache.causeway.applib.id.HasLogicalType;
import org.apache.causeway.applib.services.bookmark.BookmarkService;
import org.apache.causeway.applib.services.registry.ServiceRegistry;
import org.apache.causeway.commons.collections.ImmutableEnumSet;
import org.apache.causeway.commons.functional.Either;
import org.apache.causeway.core.config.CausewayConfiguration;
import org.apache.causeway.core.config.environment.CausewaySystemEnvironment;
import org.apache.causeway.core.metamodel.objectmanager.ObjectManager;
import org.apache.causeway.core.metamodel.spec.ActionScope;
import org.apache.causeway.core.metamodel.spec.ObjectSpecification;
import org.apache.causeway.core.metamodel.specloader.SpecificationLoader;
import org.apache.causeway.viewer.graphql.model.registry.GraphQLTypeRegistry;
import org.apache.causeway.viewer.graphql.model.types.TypeMapper;

import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class Context {

    public final GraphQLCodeRegistry.Builder codeRegistryBuilder = GraphQLCodeRegistry.newCodeRegistry();
    public final BookmarkService bookmarkService;
    public final SpecificationLoader specificationLoader;
    public final TypeMapper typeMapper;
    public final ServiceRegistry serviceRegistry;
    public final CausewayConfiguration causewayConfiguration;
    public final CausewaySystemEnvironment causewaySystemEnvironment;
    public final ObjectManager objectManager;
    public final GraphQLTypeRegistry graphQLTypeRegistry;


    public final Map<String, GqlvDomainService> domainServiceByTypeName = new LinkedHashMap<>();
    public final Map<String, GqlvDomainObject> domainObjectByTypeName = new LinkedHashMap<>();

    public ImmutableEnumSet<ActionScope> getActionScope() {
        return causewaySystemEnvironment.getDeploymentType().isProduction()
                ? ActionScope.PRODUCTION_ONLY
                : ActionScope.ANY;
    }

    public List<ObjectSpecification> objectSpecifications() {
        return specificationLoader.snapshotSpecifications()
                .filter(x -> x.getCorrespondingClass().getPackage() != Either.class.getPackage())   // exclude the org.apache_causeway.commons.functional
                .distinct((a, b) -> a.getLogicalTypeName().equals(b.getLogicalTypeName()))
                .filter(x -> x.isEntityOrViewModelOrAbstract() || x.getBeanSort().isManagedBeanContributing())
                .sorted(Comparator.comparing(HasLogicalType::getLogicalTypeName))
                .toList();
    }
}
