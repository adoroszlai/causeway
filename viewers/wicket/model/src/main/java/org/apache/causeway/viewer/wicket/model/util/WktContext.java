/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.viewer.wicket.model.util;

import java.util.Optional;

import org.apache.wicket.Application;
import org.apache.wicket.ThreadContext;
import org.apache.wicket.core.request.handler.ListenerRequestHandler;
import org.apache.wicket.request.cycle.RequestCycle;
import org.springframework.lang.Nullable;

import org.apache.causeway.commons.internal.base._Casts;
import org.apache.causeway.commons.internal.base._Strings;
import org.apache.causeway.core.metamodel.context.HasMetaModelContext;
import org.apache.causeway.core.metamodel.context.MetaModelContext;

import lombok.NonNull;
import lombok.experimental.UtilityClass;

/**
 * @since 2.0
 */
@UtilityClass
public class WktContext {

    @Nullable
    public String getApplicationKey() {
        final Application application = ThreadContext.getApplication(); // nullable
        return Optional.ofNullable(application)
                .map(Application::getName)
                .orElse(null);
    }

    @Nullable
    public MetaModelContext getMetaModelContext() {
        final Application application = ThreadContext.getApplication(); // nullable
        return _Casts.castTo(HasMetaModelContext.class, application)
                .map(HasMetaModelContext::getMetaModelContext)
                .orElse(null);
    }

    @Nullable
    public MetaModelContext getMetaModelContext(final @NonNull String applicationKey) {
        final Application application = Application.get(applicationKey); // nullable
        return _Casts.castTo(HasMetaModelContext.class, application)
                .map(HasMetaModelContext::getMetaModelContext)
                .orElse(null);
    }

    @Nullable
    public MetaModelContext computeIfAbsent(final MetaModelContext mmc) {
        return mmc!=null
                ? mmc
                : getMetaModelContext();
    }

    @Nullable
    public MetaModelContext computeIfAbsent(
            final @Nullable MetaModelContext mmc, final @Nullable String applicationKey) {
        return mmc!=null
                ? mmc
                : _Strings.isEmpty(applicationKey)
                        ? getMetaModelContext()
                        : getMetaModelContext(applicationKey);
    }

    public void pageReload() {
        var cycle = RequestCycle.get();
        var handler = cycle.getActiveRequestHandler();
        if(handler instanceof ListenerRequestHandler) {
            var currentPage = ((ListenerRequestHandler)handler).getPage();
            cycle.setResponsePage(currentPage);
        }
    }

}
