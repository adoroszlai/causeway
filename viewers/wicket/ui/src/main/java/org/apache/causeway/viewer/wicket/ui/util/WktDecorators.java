/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.viewer.wicket.ui.util;

import java.util.Optional;

import org.apache.wicket.Component;
import org.apache.wicket.ajax.markup.html.AjaxLink;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.form.Button;
import org.apache.wicket.markup.html.list.ListItem;

import org.apache.causeway.applib.fa.FontAwesomeLayers;
import org.apache.causeway.commons.internal.base._Strings;
import org.apache.causeway.core.metamodel.object.ManagedObject;
import org.apache.causeway.core.metamodel.spec.feature.ObjectAction;
import org.apache.causeway.viewer.commons.model.decorators.ActionDecorators.ActionDecorationModel;
import org.apache.causeway.viewer.commons.model.decorators.ConfirmDecorator;
import org.apache.causeway.viewer.commons.model.decorators.ConfirmDecorator.ConfirmDecorationModel;
import org.apache.causeway.viewer.commons.model.decorators.DangerDecorator;
import org.apache.causeway.viewer.commons.model.decorators.DisablingDecorator;
import org.apache.causeway.viewer.commons.model.decorators.FormLabelDecorator;
import org.apache.causeway.viewer.commons.model.decorators.IconDecorator;
import org.apache.causeway.viewer.commons.model.decorators.PrototypingDecorator;
import org.apache.causeway.viewer.commons.model.decorators.TooltipDecorator;
import org.apache.causeway.viewer.commons.model.decorators.TooltipDecorator.TooltipDecorationModel;
import org.apache.causeway.viewer.commons.model.layout.UiPlacementDirection;
import org.apache.causeway.viewer.wicket.model.links.LinkAndLabel;
import org.apache.causeway.viewer.wicket.model.links.Menuable;
import org.apache.causeway.viewer.wicket.ui.components.actionmenu.FontAwesomeBehavior;
import org.apache.causeway.viewer.wicket.ui.components.widgets.linkandlabel.ActionLink;
import org.apache.causeway.viewer.wicket.ui.util.BootstrapConstants.ButtonSemantics;

import lombok.Getter;
import lombok.experimental.Accessors;
import lombok.experimental.UtilityClass;

import de.agilecoders.wicket.core.markup.html.bootstrap.components.TooltipConfig.Placement;
import de.agilecoders.wicket.extensions.markup.html.bootstrap.confirmation.ConfirmationConfig;

@UtilityClass
public class WktDecorators {

    // -- BASIC DECORATORS
    @Getter(lazy = true) @Accessors(fluent=true) private final static FormLabel formLabel = new FormLabel();
    @Getter(lazy = true) @Accessors(fluent=true) private final static Tooltip tooltip = new Tooltip();
    @Getter(lazy = true) @Accessors(fluent=true) private final static Disable disable = new Disable();
    @Getter(lazy = true) @Accessors(fluent=true) private final static Prototyping prototyping = new Prototyping();
    @Getter(lazy = true) @Accessors(fluent=true) private final static Confirm confirm = new Confirm();
    @Getter(lazy = true) @Accessors(fluent=true) private final static Danger danger = new Danger();
    @Getter(lazy = true) @Accessors(fluent=true) private final static FaIcon faIcon = new FaIcon();
    @Getter(lazy = true) @Accessors(fluent=true) private final static MissingIcon missingIcon = new MissingIcon();

    // -- BASIC DECORATOR CLASSES

    public final static class FormLabel implements FormLabelDecorator<Component> {
        @Override
        public void decorate(final Component uiComponent, final FormLabelDecorationModel decorationModel) {
            if(decorationModel.isMandatoryIndicatorShown()) {
                Wkt.cssAppend(uiComponent, "mandatory");
            }
            uiComponent.setEscapeModelStrings(true);
        }
    }

    public final static class Tooltip implements TooltipDecorator<Component> {
        @Override
        public void decorate(final Component uiComponent, final TooltipDecorationModel decorationModel) {
            WktTooltips.addTooltip(uiComponent, decorationModel);
        }
    }

    public final static class Disable implements DisablingDecorator<Component> {
        @Override
        public void decorate(final Component uiComponent, final DisablingDecorationModel decorationModel) {
            var tooltipDecorationModel = TooltipDecorationModel
                    .ofBody(UiPlacementDirection.BOTTOM, decorationModel.reason());
            tooltip().decorate(uiComponent, tooltipDecorationModel);

            Wkt.cssAppend(uiComponent, "disabled");
            uiComponent.setEnabled(false);
        }
    }

    public final static class Prototyping implements PrototypingDecorator<Component, Component> {
        @Override
        public Component decorate(final Component uiComponent, final PrototypingDecorationModel decorationModel) {
            Wkt.cssAppend(uiComponent, "prototype");
            return uiComponent;
        }
    }

    public final static class Confirm implements ConfirmDecorator<Component> {
        @Override
        public void decorate(final Component uiComponent, final ConfirmDecorationModel decorationModel) {
            var confirmationConfig = new ConfirmationConfig()
                    .withTitle(decorationModel.title())
                    .withBtnOkLabel(decorationModel.okLabel())
                    .withBtnCancelLabel(decorationModel.cancelLabel())
                    .withBtnOkClass(ButtonSemantics.DANGER.fullButtonCss())
                    .withBtnCancelClass(ButtonSemantics.SECONDARY.fullButtonCss())
                    .withPlacement(Placement.valueOf(decorationModel.placement().name().toLowerCase()));

            Wkt.behaviorAddConfirm(uiComponent, confirmationConfig);

            if(uiComponent instanceof Button) {
                // ensure dialog ok buttons receive the danger style as well
                // don't care if expressed twice
                WktDecorators.danger().decorate(uiComponent);
            }
        }
    }

    public final static class Danger implements DangerDecorator<Component> {
        @Override
        public void decorate(final Component uiComponent) {
            Wkt.cssAppend(uiComponent, ButtonSemantics.DANGER.buttonDefaultCss());
        }
    }

    public final static class FaIcon implements IconDecorator<Component, Component> {
        @Override
        public Component decorate(final Component uiComponent, final Optional<FontAwesomeLayers> faLayers) {
            if(faLayers.isPresent()) {
                uiComponent.add(new FontAwesomeBehavior(faLayers.get()));
            }
            return uiComponent;
        }
    }

    public final static class MissingIcon implements IconDecorator<Component, Component> {
        @Override
        public Component decorate(final Component uiComponent, final Optional<FontAwesomeLayers> faLayers) {
            if(faLayers.isEmpty()) {
                Wkt.cssAppend(uiComponent, "menuLinkSpacer");
            }
            return uiComponent;
        }
    }

    // -- ACTION DECORATION
    
    public void decorateMenuAction(
            ListItem<? extends Menuable> listItem, 
            AjaxLink<ManagedObject> actionLink,
            Label actionLabel,
            ActionDecorationModel decorationModel) {
        
        decorationModel.disabling()
            .ifPresentOrElse(disableUiModel->{
                //DISABLE
                disable().decorate(listItem, disableUiModel);
                tooltip().decorate(listItem,
                        TooltipDecorationModel.ofBody(UiPlacementDirection.BOTTOM, disableUiModel.reason()));
            }, ()->{
                //DESCRIBE
                decorationModel
                    .describedAs()
                    .ifPresent(describedAs->
                        tooltip()
                            .decorate(listItem,
                                    TooltipDecorationModel.ofBody(UiPlacementDirection.BOTTOM, describedAs)));
                
                //{CAUSEWAY-1626] confirmation dialog for no-parameter menu actions
                if (decorationModel.isImmediateConfirmationRequired()) {
                    var confirmUiModel = ConfirmDecorationModel.areYouSure(UiPlacementDirection.BOTTOM);
                    confirm().decorate(actionLink, confirmUiModel);
                }
                
            });
        
        //PROTOTYPING
        decorationModel.prototyping().ifPresent(protoDecModel->{
            prototyping().decorate(actionLink, protoDecModel);
        });
        
        //CSS
        Wkt.cssAppend(listItem, decorationModel.featureIdentifier());
        decorationModel.additionalCssClass()
            .ifPresent(cssClass->Wkt.cssAppend(actionLink, cssClass));
        
        //FA-ICON
        var faLayers = decorationModel.fontAwesomeLayers();
        faIcon().decorate(actionLabel, faLayers);
        missingIcon().decorate(actionLink, faLayers);
    }

    /**
     * For rendering actions within additional-link panels or drop-downs.
     */
    public void decorateAdditionalLink(
            AjaxLink<ManagedObject> actionLink,
            Label actionLabel, 
            ActionDecorationModel decorationModel) {

        decorationModel.disabling()
            .ifPresentOrElse(disableUiModel->{
                //DISABLE
                disable().decorate(actionLink, disableUiModel);
            }, ()->{
                //DESCRIBE
                decorationModel
                    .describedAs()
                    .ifPresent(describedAs->WktTooltips.addTooltip(actionLink, describedAs));
            });

        if (ObjectAction.Util.returnsBlobOrClob(decorationModel.action())) {
            Wkt.cssAppend(actionLink, "noVeil");           
        }
        
        //PROTOTYPING
        decorationModel.prototyping().ifPresent(protoDecModel->{
            prototyping().decorate(actionLink, protoDecModel);
        });
    
        //CSS
        Wkt.cssAppend(actionLink, decorationModel.featureIdentifier());
        decorationModel.additionalCssClass()
            .ifPresent(cssClass->Wkt.cssAppend(actionLink, cssClass));
    
        //FA-ICON
        var faLayers = decorationModel.fontAwesomeLayers();
        faIcon().decorate(actionLabel, faLayers);
    }

    //TODO[CAUSEWAY-3824] WIP make more generic
    public void decorateAdditionalLink(
            final LinkAndLabel linkAndLabel, 
            final Component tooltipReceiver, 
            final Label viewTitleLabel,
            final boolean isForceAlignmentWithBlankIcon,
            final ActionDecorationModel decorationModel) {
        
        var link = linkAndLabel.getUiComponent();
        var action = linkAndLabel.getManagedAction().getAction();
        var hasDisabledReason = link instanceof ActionLink 
                && _Strings.isNotEmpty(((ActionLink) link).getReasonDisabledIfAny());
        
        WktTooltips.addTooltip(tooltipReceiver, hasDisabledReason
                ? ((ActionLink) link).getReasonDisabledIfAny()
                : linkAndLabel.getDescription().orElse(null));
        
        if(ObjectAction.Util.returnsBlobOrClob(action)) {
            Wkt.cssAppend(link, "noVeil");
        }
        if(action.isPrototype()) {
            Wkt.cssAppend(link, "prototype");
        }
        Wkt.cssAppend(link, linkAndLabel.getFeatureIdentifier());
        
        if (action.getSemantics().isAreYouSure()) {
            if(action.getParameterCount()==0) {
                if (!hasDisabledReason) {
                    var confirmUiModel = ConfirmDecorationModel
                            .areYouSure(UiPlacementDirection.BOTTOM);
                    WktDecorators.confirm().decorate(link, confirmUiModel);
                }
            }
            // ensure links receive the danger style
            // don't care if expressed twice
            WktDecorators.danger().decorate(link);
        } else {
            Wkt.cssAppend(link, linkAndLabel.isRenderOutlined()
                    || action.isPrototype()
                    ? ButtonSemantics.SECONDARY.buttonOutlineCss()
                    : ButtonSemantics.SECONDARY.buttonDefaultCss());
        }
        
        linkAndLabel
            .getAdditionalCssClass()
            .ifPresent(cssClass->Wkt.cssAppend(link, cssClass));
        
        var faLayers = linkAndLabel.lookupFontAwesomeLayers(isForceAlignmentWithBlankIcon);
        WktDecorators.faIcon().decorate(viewTitleLabel, faLayers);
        WktDecorators.missingIcon().decorate(viewTitleLabel, faLayers);
    }

}
