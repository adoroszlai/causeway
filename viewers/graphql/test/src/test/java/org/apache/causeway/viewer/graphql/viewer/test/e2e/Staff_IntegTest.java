/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.viewer.graphql.viewer.test.e2e;

import java.util.Optional;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Propagation;

import org.apache.causeway.applib.services.bookmark.Bookmark;
import org.apache.causeway.commons.internal.collections._Maps;
import org.apache.causeway.viewer.graphql.viewer.test.domain.dept.Department;
import org.apache.causeway.viewer.graphql.viewer.test.domain.dept.StaffMember;

import lombok.val;


//NOT USING @Transactional since we are running server within same transaction otherwise
@Order(60)
@ActiveProfiles("test")
public class Staff_IntegTest extends Abstract_IntegTest {

    @Test
    @UseReporter(DiffReporter.class)
    void list_all_staff_members() throws Exception {

        // when, then
        Approvals.verify(submit(), jsonOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void find_staff_member_by_name_and_edit_grade_choices() throws Exception {

        // when, then
        Approvals.verify(submit(), jsonOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void find_staff_member_by_name_and_edit() throws Exception {

        // given
        final Optional<Bookmark> bookmarkIfAny =
                transactionService.callTransactional(
                        Propagation.REQUIRED,
                        () -> {
                            StaffMember pojo = staffMemberRepository.findByName("Gerry Jones");
                            return bookmarkService.bookmarkFor(pojo);
                        }
                ).valueAsNullableElseFail();

        assertThat(bookmarkIfAny).isPresent();

        // when, then
        Approvals.verify(submit(), jsonOptions());

        // and in the database...
        final Optional<StaffMember> staffMemberIfAny =
                transactionService.callTransactional(
                        Propagation.REQUIRED,
                        () -> bookmarkService.lookup(bookmarkIfAny.get(), StaffMember.class)
                ).valueAsNullableElseFail();

        assertThat(staffMemberIfAny).isPresent();
        assertThat(staffMemberIfAny.get()).extracting(StaffMember::getName).isEqualTo("Gerald Johns");
    }

    @Test
    @UseReporter(DiffReporter.class)
    void staff_member_name_validate() throws Exception {

        // when, then
        Approvals.verify(submit(), jsonOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void staff_member_name_edit_invalid() throws Exception {

        // when, then
        Approvals.verify(submit(), jsonOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void create_staff_member_with_department() throws Exception {

        final Bookmark bookmark =
                transactionService.callTransactional(
                        Propagation.REQUIRED,
                        () -> {
                            Department department = departmentRepository.findByName("Classics");
                            return bookmarkService.bookmarkFor(department).orElseThrow();
                        }
                ).valueAsNonNullElseFail();

        val response = submit(_Maps.unmodifiable("$departmentId", bookmark.getIdentifier()));

        // then payload
        Approvals.verify(response, jsonOptions());

    }
}
