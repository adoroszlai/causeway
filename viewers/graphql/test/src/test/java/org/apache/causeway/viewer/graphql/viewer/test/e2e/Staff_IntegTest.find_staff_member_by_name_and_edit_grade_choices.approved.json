{
  "data" : {
    "university_dept_Staff" : {
      "findByName" : {
        "invoke" : {
          "grade" : {
            "disabled" : null,
            "choices" : [ "PROFESSOR", "READER", "SENIOR_LECTURER", "LECTURER", "RESEARCH_ASSISTANT" ]
          }
        }
      }
    }
  }
}